### PR TITLE
Refactoring + Tests + MSI 100%

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

**Review by samdark — APPROVED** [review]

File: src/FileRouter.php [hunk 1]
```diff
@@ -155,61 +155,57 @@ private function parseRequestPath(ServerRequestInterface $request): Generator
...
         );
 
-        if (!preg_match('#^(.*?)/([^/]+)/?$#', $controllerName, $matches)) {
+        if (!preg_match('#^/?(.*?)/([^/]+)/?$#', $controllerName, $matches)) {
             return;
         }
 
-        $directoryPath = $matches[1];
-        $controllerName = $matches[2];
+        [$_, $directoryPath, $controllerName] = $matches;
```
Comment (xepozz) [inline src/FileRouter.php]: I'd leave it as it was before

File: src/FileRouter.php [hunk 2]
```diff
@@ -155,61 +155,57 @@ private function parseRequestPath(ServerRequestInterface $request): Generator
...
         $controllerName = preg_replace_callback(
-            '#(/.)#u',
-            static fn(array $matches) => mb_strtoupper($matches[1]),
+            '#/.#u',
+            static fn(array $matches) => mb_strtoupper($matches[0]),
             $path,
         );
 
-        if (!preg_match('#^(.*?)/([^/]+)/?$#', $controllerName, $matches)) {
+        if (!preg_match('#^/?(.*?)/([^/]+)/?$#', $controllerName, $matches)) {
```
Comment (xepozz) [inline src/FileRouter.php]: Why?

File: src/FileRouter.php [hunk 3]
```diff
@@ -155,61 +155,57 @@ private function parseRequestPath(ServerRequestInterface $request): Generator
...
-                $this->namespace . '\\' . $this->baseControllerDirectory . '\\' . $directoryPath . '\\' . $controllerName . $this->classPostfix
-            ),
-            $possibleAction,
-        ];
     }
 
-    private function cleanClassname(string $className): string
+    private function makeClassName(string $controllerName, string $directoryPath = ''): string
     {
-        return str_replace(
```
Comment (xepozz) [inline src/FileRouter.php:209]: As far as I remember there were some Windows cases where such operations are necessary to make the router work correctly

File: src/FileRouter.php [hunk 4]
```diff
@@ -155,61 +155,57 @@ private function parseRequestPath(ServerRequestInterface $request): Generator
...
-        yield [
-            $this->cleanClassname(
-                $this->namespace . '\\' . $this->baseControllerDirectory . '\\' . $directoryPath . '\\' . $controllerName . $this->classPostfix
-            ),
-            $possibleAction,
-        ];
     }
 
-    private function cleanClassname(string $className): string
+    private function makeClassName(string $controllerName, string $directoryPath = ''): string
```
Comment (xepozz) [inline src/FileRouter.php]: No need in optional parameters, pass default values directly

**Review by xepozz — CHANGES_REQUESTED** [review]

File: src/FileRouter.php (hunk 2 above)
Comment (samdark) [reply src/FileRouter.php]: That enables `/` prefixing controller name. Use-case for that is interesting.

File: src/FileRouter.php (hunk 1 above)
Comment (vjik) [reply src/FileRouter.php]: Replaced to `[, $directoryPath, $controllerName] = $matches;`

File: src/FileRouter.php (hunk 2 above)
Comment (vjik) [reply src/FileRouter.php]: Here `$controllerName` contains path with uppercased first symbols. Renamed it to `$pathAsNamespace`.

This regexp should be valid for paths with one and more items (`/user`, `/user/profile`, `/user/profile/view`).

File: src/FileRouter.php (hunk 4 above)
Comment (vjik) [reply src/FileRouter.php]: done

File: src/FileRouter.php (hunk 3 above)
Comment (vjik) [reply src/FileRouter.php:209]: Seems, path always use `/`. See https://github.com/php-fig/http-message/blob/402d35bcb92c70c026d1a6a9883f06b2ead23d71/src/UriInterface.php#L135

**Review by samdark — APPROVED** [review]

**Pull request merged**